tests/test_pipmanager.py

```diff
@@ -110,3 +112,38 @@ def test_say_hi_on_first_install(self):
             logassert.setup(self, 'fades.pipmanager')
             mgr.install('bar')
             self.assertNotLoggedInfo("Hi! This is fades")
+
+    def test_brute_force_install_pip_installer_exists(self):
+        mgr = PipManager('/usr/bin', pip_installed=False)
+        with patch.object(helpers, 'logged_exec') as mocked_exec:
+            with patch('os.path.exists') as os_exists:
+                os_exists.return_value = True
+                mgr.brute_force_install_pip()
+
+                os_exists.assert_called_once_with(mgr.pip_installer_fname)
+            mocked_exec.assert_called_with(['/usr/bin/python', mgr.pip_installer_fname, '-I'])
+        self.assertTrue(mgr.pip_installed)
+
+    def test_brute_force_install_pip_no_installer(self):
+        mgr = PipManager('/usr/bin', pip_installed=False)
+        with patch.object(helpers, 'logged_exec') as mocked_exec:
+            with patch('os.path.exists') as os_exists:
+                with patch.object(mgr, 'download_pip_installer') as download_installer:
+                    os_exists.return_value = False
+                    mgr.brute_force_install_pip()
+
+                    os_exists.assert_called_once_with(mgr.pip_installer_fname)
+                    download_installer.assert_called_once_with()
+            mocked_exec.assert_called_with(['/usr/bin/python', mgr.pip_installer_fname, '-I'])
+        self.assertTrue(mgr.pip_installed)
+
+    def test_download_pip_installer(self):
+        mgr = PipManager('/usr/bin', pip_installed=False)
+        with patch('fades.pipmanager.request.urlopen') as urlopen, \
+                patch('shutil.copy') as copy, \
+                patch('shutil.copyfileobj') as copyobj:
+            mgr.download_pip_installer()
+
+        urlopen.assert_called_once_with(pipmanager.PIP_INSTALLER)
+        self.assertEqual(copyobj.call_count, 1)
+        copy.assert_has_calls([mock.call(mock.ANY, mgr.pip_installer_fname)])
```

Comment:
Generally, for the whole tests (let's tallk IRL a little about this)...

I don't like patching those low level stdlib calls... for example, patching os.path.exists, it's a kind of mess if the tests infrastructure itself uses (now or in the future) that call while you have it patched.

For these kind of tests, what I would do is to patch the filename that the code will use to make it use some file in a temporary dict. Then, you could try the code if the file is there, if the file is not there, etc. The only thing I would patch is request.urlopen (we can be sure that the unittest infrastructure will not hit the network!).

The only test that may be tricky to do is to simulate the intermediate failure when downloading the code. However, I think that if you patch request.urlopen and make it raise an exception, it would fit for this (the file will not have some stuff in it, but it will be created (even if empty), and it will simulate the problem)